src/models/dirac_mixture_single_publisher_model.py

```diff
@@ -315,3 +316,161 @@ def predict(
             for c in self.components
         ]
         return sum([w * pmf for w, pmf in zip(self.ws, scaled_component_pmfs)])
+
+
+class DiracMixtureSinglePublisherModel(ReachCurve):
+    def __init__(
+        self,
+        data: List[ReachPoint],
+        ncomponents: int = 200,
+    ):
+        """Constructs a Dirac mixture single publisher model.
+
+        Args:
+            data:  A list consisting of the single ReachPoint to which the model
+                is to be fit.
+            ncomponents:  Number of components in the Poisson mixture.  Based on
+                our experience, 100 components are enough, and more components
+                are also fine (they will not introduce overfitting).
+        """
+        if len(data) != 1:
+            raise ValueError("Exactly one ReachPoint must be specified")
+        if data[0].impressions[0] < 0.001:
+            raise ValueError("Attempt to create model with 0 impressions")
+        if data[0].impressions[0] < data[0].reach(1):
+            raise ValueError(
+                "Cannot have a model with fewer impressions than reached people"
+            )
+        self._data = data
+        self._reach_point = data[0]
+        self.hist = self._reach_point.zero_included_histogram
+        if data[0].spends:
+            self._cpi = data[0].spends[0] / data[0].impressions[0]
+        else:
+            self._cpi = None
+        if self._reach_point.universe_size is None:
+            raise ValueError(
+                "A fit requires the universe size to be known, "
+                "please provide a ReachPoint with a known universe size instead."
+            )
+        self.ncomponents = ncomponents
+        self._fit_computed = False
+
+    @staticmethod
+    def debiased_clip(noised_histogram: np.ndarray) -> np.ndarray:
```

Comment:
Is `noised_histogram` coming from the Halo system? If so, is `noised_histogram` already non-negative as shown in this [line](https://github.com/world-federation-of-advertisers/planning-evaluation-framework/blob/jiayu_class_dirac_mixture_single_publisher_model/src/simulator/halo_simulator.py#L199)?

Reply:
Thanks for the observation, Rieman.  [This line](https://github.com/world-federation-of-advertisers/planning-evaluation-framework/blob/jiayu_class_dirac_mixture_single_publisher_model/src/simulator/halo_simulator.py#L199) made the `kplus_reaches` nonnegative.  But the histogram here in the model is `k_reaches`: k_reaches[k] = kplus_reaches[k] - kplus_reaches[k + 1] which is not necessarily no-negative.

Reply:
`np.minimum.accumulate` in this [line](https://github.com/world-federation-of-advertisers/planning-evaluation-framework/blob/jiayu_class_dirac_mixture_single_publisher_model/src/simulator/halo_simulator.py#L199) does make sure that. For example:
```python
kplus_reaches = [100, 50, 40, 60, 20, -1, -5]
kplus_reaches = np.minimum.accumulate(np.maximum(kplus_reaches, 0))
print(kplus_reaches)
>>> [100  50  40  40  20   0   0]
```

Reply:
I see.  Thanks a lot for the explanation!!

Now I removed this `debiased_clip` function.